chips/lowrisc/src/uart.rs

```diff
@@ -0,0 +1,201 @@
+    rx_client: OptionalCell<&'a dyn hil::uart::ReceiveClient>,
+}
+
+#[derive(Copy, Clone)]
+pub struct UartParams {
+    pub baud_rate: u32,
+}
+
+impl Uart<'a> {
+    pub const fn new(base: StaticRef<UartRegisters>, clock_frequency: u32) -> Uart<'a> {
+        Uart {
+            registers: base,
+            clock_frequency: clock_frequency,
+            tx_client: OptionalCell::empty(),
+            rx_client: OptionalCell::empty(),
+        }
+    }
+
+    fn set_baud_rate(&self, baud_rate: u32) {
+        let regs = self.registers;
+        let uart_ctrl_nco = ((baud_rate as u64) << 20) / self.clock_frequency as u64;
+
+        regs.ctrl
+            .write(ctrl::nco.val((uart_ctrl_nco & 0xffff) as u32));
+        regs.ctrl.modify(ctrl::tx.val(1 as u32));
+        regs.ctrl.modify(ctrl::rx.val(1 as u32));
+    }
+}
+
+impl hil::uart::UartData<'a> for Uart<'a> {}
+impl hil::uart::Uart<'a> for Uart<'a> {}
+
+impl hil::uart::Configure for Uart<'a> {
+    fn configure(&self, params: hil::uart::Parameters) -> ReturnCode {
+        let regs = self.registers;
+        // We can set the baud rate.
+        self.set_baud_rate(params.baud_rate);
+
+        regs.fifo_ctrl.write(fifo_ctrl::rxrst.val(1 as u32));
+        regs.fifo_ctrl.modify(fifo_ctrl::txrst.val(1 as u32));
+
+        // Disable interrupts as we don't suppor them
+        regs.intr_enable.set(0);
+
+        ReturnCode::SUCCESS
+    }
+}
+
+impl hil::uart::Transmit<'a> for Uart<'a> {
+    fn set_transmit_client(&self, client: &'a dyn hil::uart::TransmitClient) {
+        self.tx_client.set(client);
+    }
+
+    fn transmit_buffer(
+        &self,
+        tx_data: &'static mut [u8],
+        tx_len: usize,
+    ) -> (ReturnCode, Option<&'static mut [u8]>) {
+        let regs = self.registers;
+
+        if tx_len == 0 {
+            return (ReturnCode::ESIZE, Some(tx_data));
+        }
+
+        while regs.status.is_set(status::txfull) {}
+
+        // Fill the TX buffer until it reports full.
+        for i in 0..tx_len {
+            // Write the byte from the array to the tx register.
+            regs.wdata.write(wdata::data.val(tx_data[i] as u32));
+
+            // Check if the buffer is full and wait until it isn't
+            while regs.status.is_set(status::txfull) {}
+        }
+
+        // Wait for the send to complete
+        while regs.status.is_set(status::txempty) != true {}
+
+        (ReturnCode::SUCCESS, None)
+    }
+
+    fn transmit_abort(&self) -> ReturnCode {
+        ReturnCode::FAIL
+    }
+
+    fn transmit_word(&self, _word: u32) -> ReturnCode {
+        ReturnCode::FAIL
+    }
+}
+
+impl hil::uart::Receive<'a> for Uart<'a> {
+    fn set_receive_client(&self, client: &'a dyn hil::uart::ReceiveClient) {
+        self.rx_client.set(client);
+    }
+
+    fn receive_buffer(
+        &self,
+        _rx_buffer: &'static mut [u8],
+        _rx_len: usize,
+    ) -> (ReturnCode, Option<&'static mut [u8]>) {
+        (ReturnCode::SUCCESS, None)
```

Comment:
Is reception currently not implemented? If so, this should return a FAIL. Please don't silently drop the request.

I'm assuming this isn't implemented because of the lack of DMA/interrupt based operations? If so, at the very least receive_word should be implemented.

Reply:
It's not implemented. I have changed it to return a failure.

It's mostly not implemented because the HiFive1 doesn't do it and I have no way to test (or no use for it).

What use case will test receiving UART?

Reply:
The process console capsule does so -- it provides a console in which you can look at the status of processes and start/stop them. If we can't easily get it working with what is currently supported, that's OK, I think we can leave reception as unimplemented for now.

Reply:
I have managed to get apps to run somewhat. I can get the user space app to run all of the assembly portion of the libtock-rs hello app. The app crashed once it jumps to the C code though. Due to this [hardware bug](https://github.com/lowRISC/opentitan/issues/1035) I am unable to debug further.

So at the moment we don't have apps working so I can't test it. I am going to mark this comment as resolved and we can re-visit implementing reception in the future.

Reply:
In imix, there is a virtual_uart_rx test, which puts UART virtualization on top of a UART receive handler and tests reception. It's a pure kernel test.